libpod/networking_linux.go

```diff
@@ -63,18 +63,65 @@ const (
 // This is need because a HostIP of 127.0.0.1 would now allow the gvproxy forwarder to reach to open ports.
 // For machine the HostIP must only be used by gvproxy and never in the VM.
 func (c *Container) convertPortMappings() []types.PortMapping {
-	if !machine.IsGvProxyBased() || len(c.config.PortMappings) == 0 {
+	if len(c.config.PortMappings) == 0 {
 		return c.config.PortMappings
 	}
-	// if we run in a machine VM we have to ignore the host IP part
+
+	var convert func(types.PortMapping) []types.PortMapping
+	if machine.IsGvProxyBased() {
+		// if we run in a qemu VM we have to ignore the host IP part
+		convert = stripHost
+	} else if machine.MachineHostType() == machine.Wsl {
+		// WSL machines do not relay ipv4 traffic to dual-stack ports, simulate instead
+		convert = splitDualStack
+	} else {
+		return c.config.PortMappings
+	}
+
 	newPorts := make([]types.PortMapping, 0, len(c.config.PortMappings))
 	for _, port := range c.config.PortMappings {
-		port.HostIP = ""
-		newPorts = append(newPorts, port)
+		newPorts = append(newPorts, convert(port)...)
 	}
+
 	return newPorts
 }
 
+func stripHost(port types.PortMapping) []types.PortMapping {
+	port.HostIP = ""
+	return []types.PortMapping{port}
+}
+
+func splitDualStack(port types.PortMapping) []types.PortMapping {
+	ports := []types.PortMapping{port}
+	protocol := port.Protocol
+	if strings.HasSuffix(protocol, "4") || strings.HasSuffix(protocol, "6") {
+		return ports
+	}
+
+	host := port.HostIP
+	if len(host) == 0 {
+		host = "0.0.0.0"
+	}
+	ip := net.ParseIP(host)
+	splitLoopback := ip.IsLoopback() && ip.To4() == nil
+	// Map ::1 and 0.0.0.0/:: to ipv4 + ipv6 to simulate dual-stack
+	if ip.IsUnspecified() || splitLoopback {
+		ports = append(ports, port)
+		ports[0].Protocol = protocol + "4"
+		ports[1].Protocol = protocol + "6"
```

Comment:
This is not correct, types.PortMapping defines Protocol as comma separated list of protocols. This will never be the case in the libpod db because we deduplicate this when we create a container. However this is a implementation detail only we should not rely on that.

Also more important using tcp4/6 is not a supported protocol. I am sure that both cni and netavark will fail with that. Should'nt it be enough to bind both `::`  and `0.0.0.0` why do we need the extra 4/6. I think this is some golang implementation detail we should not use.

Reply:
We need that protocol distinction since otherwise you would end up with a port already bound error (since they value evaluate to the same binding on Golang). I moved the code back to rootlessport, so we know it will be handled correctly.  See the reply https://github.com/containers/podman/pull/14026#issuecomment-1113755530 for more details

The comma not being handled was an oversight in moving the code (in rootlessport it was already split). Now that its moved back this should also be resolved.